lib/AST/ASTContext.cpp

```diff
@@ -2713,12 +2713,16 @@ ModuleDecl *ASTContext::getModuleByIdentifier(Identifier ModuleID) {
   return getModule(builder.get());
 }
 
-ModuleDecl *ASTContext::getLoadedModuleByABIName(StringRef ModuleName) {
-  for (auto &[_, module] : getLoadedModules()) {
-    if (ModuleName == module->getABIName().str())
-      return module;
-  }
-  return nullptr;
+llvm::SmallVector<ModuleDecl *, 1>
+ASTContext::getModulesByRealOrABIName(StringRef ModuleName) {
+  llvm::SmallVector<ModuleDecl *, 1> modules;
+
+  for (auto &[_, module] : getLoadedModules())
```

Comment:
the number of loaded modules can be large. Instead why don't you pre-populate a mapping from module name to a list of modules with that name

Reply:
Is "RealName" used elsewhere? If it's a new term, may I suggest "ASTName"?

Reply:
That was my initial idea, however the module's ABI name is mutable (there's a setter for it), so we could easily end up in a situation where keeping that mapping and the module's ABI name are out of sync...

Reply:
> Is "RealName" used elsewhere? If it's a new term, may I suggest "ASTName"?

`RealName` is an actual term already.

Reply:
> however the module's ABI name is mutable

I thought this gets set via `-module-abi-name` and is global? Is there a legitimate use-case for setting the ABI name in the middle of a compilation?

Reply:
Yes it's set via `-module-abi-name` and global, what I mean is, on `ModuleDecl` there's a `setABIName` method which is called at some point by the module loader/frontend. I can implement a working solution with the mapping now, but my worry is that in the future it could be quite easy to add a call to `setABIName` somewhere else and make it so the mapping is out of sync.

Reply:
Actually it looks like the `ModuleDecl` has access to the AST it's in, so I should be able to do this in a way that avoids problems in the future.